R/count_labels.R

```diff
@@ -0,0 +1,211 @@
+  }
+
+  ns <- NS(inputId)
+  counttotal <- sum(countsmax)
+
+  lapply(seq_along(choices), function(i) {
+    choice <- as.character(choices[i])
+    countmax <- countsmax[i]
+    countnow <- if (is.null(countsnow)) 0 else countsnow[i]
+
+    countBar(
+      inputId = ns(i),
+      label = choice,
+      countmax = countmax,
+      countnow = countnow,
+      counttotal = counttotal
+    )
+  })
+}
+
+#' Progressbar with label
+#'
+#' Progress bar with label
+#' @param session (`session`) object passed to function given to `shinyServer`.
+#' @param inputId (`character(1)`) `shiny` id
+#' @param label (`character(1)`) Text to display followed by counts
+#' @param countmax (`numeric(1)`) maximal possible count for a single item.
+#' @param countnow (`numeric(1)`) current count of a single item.
+#' @param counttotal (`numeric(1)`) total count to make whole progress bar
+#'  taking part of the container. Ratio between `countmax / counttotal`
+#'  determines `<style="width: <countmax / counttotal>%""`.
+#' @return `shiny.tag` object with a progress bar and a label.
+#' @keywords internal
+countBar <- function(inputId, label, countmax, countnow = NULL, counttotal = countmax) {
+  checkmate::assert_string(inputId)
+  checkmate::assert_string(label)
+  checkmate::assert_number(countmax)
+  checkmate::assert_number(countnow, null.ok = TRUE, upper = countmax)
+  checkmate::assert_number(counttotal, lower = countmax)
+
+  label <- make_count_text(label, countmax = countmax, countnow = countnow)
+  ns <- NS(inputId)
+  tags$div(
+    class = "progress state-count-container",
+    # * .9 to not exceed width of the parent html element
+    tags$div(
+      id = ns("count_bar_filtered"),
+      class = "progress-bar state-count-bar-filtered",
+      style = sprintf("width: %s%%", countnow / counttotal * 100),
+      role = "progressbar",
+      label
+    ),
+    tags$div(
+      id = ns("count_bar_unfiltered"),
+      class = "progress-bar state-count-bar-unfiltered",
+      style = sprintf("width: %s%%", (countmax - countnow) / counttotal * 100),
+      role = "progressbar"
+    )
+  )
+}
+
+#' @rdname countBars
+updateCountBars <-  function(session = getDefaultReactiveDomain(), inputId, choices,
+                                  countsmax, countsnow = NULL) {
+    checkmate::assert_string(inputId)
+    checkmate::assert_vector(choices)
+    checkmate::assert_numeric(countsmax, len = length(choices))
+    checkmate::assert_numeric(countsnow, len = length(choices), null.ok = TRUE)
+
+    ns <- NS(inputId)
+    counttotal <- sum(countsmax)
+    lapply(seq_along(choices), function(i) {
+      choice <- choices[i]
+      countmax <- countsmax[i]
+      countnow <- if (is.null(countsnow)) countmax else countsnow[i]
+      updateCountBar(
+        inputId = ns(i),
+        label = choice,
+        countmax = countmax,
+        countnow = countnow,
+        counttotal = counttotal
+      )
+    })
+  invisible(NULL)
+}
```

Comment:
Also suggest `mapply` here

```suggestion
updateCountBars <-  function(session = getDefaultReactiveDomain(), inputId, choices,
                                  countsmax, countsnow = NULL) {
    checkmate::assert_string(inputId)
    checkmate::assert_vector(choices)
    checkmate::assert_numeric(countsmax, len = length(choices))
    checkmate::assert_numeric(countsnow, len = length(choices), null.ok = TRUE)
    if (!is.null(countsnow)) {
      checkmate::assert_true(all(countsnow <= countsmax))
    } else {
      countsnow <- rep(0L, length(choices))
    }

    ns <- NS(inputId)
    mapply(
      updateCountBar,
      inputId = ns(seq_along(choices)),
      label = choices,
      countmax = countsmax,
      countnow = countsnow,
      MoreArgs = list(
        counttotal = sum(countsmax)
      )
    )
  invisible(NULL)
}
```